pkg/services/annotations/annotationsimpl/loki/historian_store_test.go

```diff
@@ -0,0 +1,105 @@
+package loki
+
+import (
+	"testing"
+
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/stretchr/testify/require"
+)
+
+func TestUseStore(t *testing.T) {
+	t.Run("false if state history disabled", func(t *testing.T) {
+		cfg := setting.UnifiedAlertingStateHistorySettings{
+			Enabled: false,
+		}
+		use := useStore(cfg, featuremgmt.WithFeatures())
+		require.False(t, use)
+	})
+
+	t.Run("false if any invalid backend", func(t *testing.T) {
+		t.Run("single", func(t *testing.T) {
+			cfg := setting.UnifiedAlertingStateHistorySettings{
+				Enabled: true,
+				Backend: "invalid-backend",
+			}
+			use := useStore(cfg, featuremgmt.WithFeatures())
+			require.False(t, use)
+		})
+
+		t.Run("primary", func(t *testing.T) {
+			cfg := setting.UnifiedAlertingStateHistorySettings{
+				Enabled:      true,
+				Backend:      "multiple",
+				MultiPrimary: "invalid-backend",
+			}
+			use := useStore(cfg, featuremgmt.WithFeatures())
+			require.False(t, use)
+		})
+
+		t.Run("secondary", func(t *testing.T) {
+			cfg := setting.UnifiedAlertingStateHistorySettings{
+				Enabled:          true,
+				Backend:          "multiple",
+				MultiPrimary:     "annotations",
+				MultiSecondaries: []string{"annotations", "invalid-backend"},
+			}
+			use := useStore(cfg, featuremgmt.WithFeatures())
+			require.False(t, use)
+		})
+	})
+
+	t.Run("false if no backend is Loki", func(t *testing.T) {
+		cfg := setting.UnifiedAlertingStateHistorySettings{
+			Enabled: true,
+			Backend: "annotations",
+		}
+		use := useStore(cfg, featuremgmt.WithFeatures())
+		require.False(t, use)
+	})
+
+	t.Run("true if any backend is Loki", func(t *testing.T) {
+		t.Run("single", func(t *testing.T) {
+			cfg := setting.UnifiedAlertingStateHistorySettings{
+				Enabled: true,
+				Backend: "loki",
+			}
+			features := featuremgmt.WithFeatures(
+				featuremgmt.FlagAlertStateHistoryLokiOnly,
+				featuremgmt.FlagAlertStateHistoryLokiPrimary,
+				featuremgmt.FlagAlertStateHistoryLokiSecondary,
+			)
+			use := useStore(cfg, features)
+			require.True(t, use)
+		})
+
+		t.Run("primary", func(t *testing.T) {
```

Comment:
Something I had not considered...

In this mode, we are dual-writing to both Loki and Annotations. So, *both* an annotation and a loki entry are created on every event.

With how the composite store works, that means we'd see duplicated annotations in the graph? We could consider disabling this for the dual write mode. I.e., only enabled if LokiOnly is also enabled.

Reply:
That's a good point, the API would return duplicates and I'm not sure how the frontend would handle that. It'd be better to just not bother querying Loki if these entries already exist in annotations.